### PR TITLE
Rework on Stack Array Data Structure

### PR DESCRIPTION
**Pull request by TruongNhanNguyen**
What I have done in this PR:
- Reorganize code in `data_structure` namespace and `Stack` class
- The destructor is added to the `Stack` class to ensure memory deallocation
- Comments are added to the member functions to describe their purpose
- The self-test function is provided, which demonstrates the usage of the `Stack` class and its methods.

#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes:

### CODE REVIEW

**Comment by TruongNhanNguyen** [comment]
Please take a look on this PR @realstealthninja, @Panquesito7

**Comment by TruongNhanNguyen** [comment]
Could you have a look on this PR @realstealthninja

**Review by mrmagic2020 — APPROVED** [review]
Both maintainers of this repo are rather busy these days, so I'll do a quick review. 
Very solid rework and documentation. Have you considered adding methods like `empty()` and `full()`? They might help client code to handle these cases before operations like `pop()` and `push()`.
Overall I think this is merge-worthy. Up to the maintainers to make further decision :)

**Comment by mrmagic2020** [comment]
> Can you add **don't-close** label for this PR, @mrmagic2020?

Sorry man, I don't have write access to this repo... you still need to wait for the maintainers

**Review by mrmagic2020 — APPROVED** [review]

File: data_structures/stack_using_array.cpp [hunk 1]
```diff
@@ -1,62 +1,178 @@
...
-        std::cout << "\n" << stack[--stack_idx] << " deleted";
+    /**
+     * @brief Pops an element from the stack
+     *
+     * @return The popped element
+     * @throws std::out_of_range if the stack is empty
+     */
+    T pop() {
+        if (empty()) {
+            throw std::out_of_range("Stack is empty");
```
Comment (realstealthninja) [inline data_structures/stack_using_array.cpp:81]:
```suggestion
            throw std::out_of_range("Stack underflow");
```

File: data_structures/stack_using_array.cpp [hunk 2]
```diff
@@ -1,62 +1,178 @@
...
+    assert(stack.empty() == true);
+    assert(stack.full() == false);
+
+    assert(stack.push(10) == true);
+    assert(stack.push(20) == true);
+    assert(stack.push(30) == true);
+    assert(stack.push(40) == true);
+    assert(stack.push(50) == true);
+    assert(stack.push(60) == false);
+
```
Comment (realstealthninja) [inline data_structures/stack_using_array.cpp:143]: Add tests for overflow and underflow

File: data_structures/stack_using_array.cpp [hunk 3]
```diff
@@ -1,62 +1,178 @@
...
+     * @return true if the element was successfully pushed onto the stack, false
+     * otherwise
+     */
+    bool push(T element) {
+        if (full()) {
+            return false;
+        } else {
+            stack[stackIndex++] = element;
+            return true;
+        }
```
Comment (realstealthninja) [inline data_structures/stack_using_array.cpp:70]:
```suggestion
    void push(T element) {
        if (full()) {
            throw std::out_of_range("Stack overflow");
        } else {
            stack[stackIndex++] = element;
        }
```

File: data_structures/stack_using_array.cpp [hunk 4]
```diff
@@ -1,62 +1,178 @@
...
+    int stackSize;   ///< Maximum size of the stack
+    int stackIndex;  ///< Index pointing to the top element of the stack
+
+ public:
+    /**
+     * @brief Constructs a new Stack object
+     *
+     * @param size Maximum size of the stack
+     */
+    Stack(int size) : stackSize(size), stackIndex(0) { stack = new T[size]; }
```
Comment (realstealthninja) [inline data_structures/stack_using_array.cpp:36]:
```suggestion
    Stack(int size) : stackSize(size), stackIndex(-1) { stack = new T[size]; }
```
TOS should be -1 at the start

File: data_structures/stack_using_array.cpp [hunk 5]
```diff
@@ -1,62 +1,178 @@
...
+     * @return true if the stack is full, false otherwise
+     */
+    bool full() const { return stackIndex == stackSize; }
+
+    /**
+     * @brief Checks if the stack is empty
+     *
+     * @return true if the stack is empty, false otherwise
+     */
+    bool empty() const { return stackIndex == 0; }
```
Comment (realstealthninja) [inline data_structures/stack_using_array.cpp:55]:
```suggestion
    bool empty() const { return stackIndex == -1; }
```

File: data_structures/stack_using_array.cpp [hunk 6]
```diff
@@ -1,62 +1,178 @@
...
+ * @returns void
+ */
+static void test() {
+    data_structures::Stack<int> stack(5);
+
+    // Test push, pop, topmost, bottom, full, and empty operations
+    assert(stack.empty() == true);
+    assert(stack.full() == false);
+
+    assert(stack.push(10) == true);
```
Comment (realstealthninja) [inline data_structures/stack_using_array.cpp:137]: Since we are not returning true any use this as a guide

```suggestion
    stack.push(10);
    assert(stack.topmost() == 10);
```

**Review by realstealthninja — CHANGES_REQUESTED** [review]
Very lovely contribution. Other than some technical assumptions being wrong. This is a great implementation. I would recommend using TOS instead of stack index. The bottom is unnecessary imo since stack follows lifo principle

**Comment by TruongNhanNguyen** [comment]
Thanks, @realstealthninja. Can you have a look?

File: data_structures/stack_using_array.cpp [hunk 7]
```diff
@@ -1,62 +1,176 @@
...
+
+/**
+ * @brief Self-test implementations
+ * @returns void
+ */
+static void test() {
+    data_structures::Stack<int> stack(5);
+
+    // Test empty and full operations
+    assert(stack.empty() == true);
```
Comment (realstealthninja) [inline data_structures/stack_using_array.cpp:111]:
```suggestion
    assert(stack.empty());
```
assert checks for booleans so this should work by default

File: data_structures/stack_using_array.cpp [hunk 8]
```diff
@@ -1,62 +1,176 @@
...
+    assert(stack.full() == false);
```
Comment (realstealthninja) [inline data_structures/stack_using_array.cpp:112]:
```suggestion
    assert(!stack.full());
```

File: data_structures/stack_using_array.cpp [hunk 9]
```diff
@@ -1,62 +1,176 @@
...
+    assert(stack.pop() == 30);
+
+    // Check topmost and bottom elements
+    assert(stack.topmost() == 20);
+    assert(stack.bottom() == 10);
+
+    assert(stack.pop() == 20);
+    assert(stack.pop() == 10);
+
+    assert(stack.empty() == true);
```
Comment (realstealthninja) [inline data_structures/stack_using_array.cpp:145]:
```suggestion
    assert(stack.empty());
```

File: data_structures/stack_using_array.cpp [hunk 10]
```diff
@@ -1,62 +1,176 @@
...
+    stack.push(10);
+    assert(stack.topmost() == 10);
 
+    stack.push(20);
+    assert(stack.topmost() == 20);
+
+    stack.push(30);
+    stack.push(40);
+    stack.push(50);
+    assert(stack.full() == true);
```
Comment (realstealthninja) [inline data_structures/stack_using_array.cpp:124]:
```suggestion
    assert(stack.full());
```

File: data_structures/stack_using_array.cpp [hunk 11]
```diff
@@ -1,62 +1,176 @@
...
+
+    // Check topmost and bottom elements
+    assert(stack.topmost() == 20);
+    assert(stack.bottom() == 10);
+
+    assert(stack.pop() == 20);
+    assert(stack.pop() == 10);
+
+    assert(stack.empty() == true);
+    assert(stack.full() == false);
```
Comment (realstealthninja) [inline data_structures/stack_using_array.cpp:146]:
```suggestion
    assert(!stack.full());
```

**Review by realstealthninja — CHANGES_REQUESTED** [review]

**Comment by TruongNhanNguyen** [comment]
Thanks for the quick fixes.

**Comment by realstealthninja** [comment]
@TruongNhanNguyen could you delete the stack.hpp file from that directory as well? thanks.

**Comment by realstealthninja** [comment]
On second thought revert that, we can do that in another pr. Sorry for the confusion. The stack.hpp problem seems out of the scope of this pr

File: data_structures/stack_using_array.cpp [hunk 12]
```diff
@@ -1,62 +1,176 @@
+#include <cassert>
 #include <iostream>
+#include <memory>
+#include <stdexcept>
```
Comment (realstealthninja) [inline data_structures/stack_using_array.cpp:4]:
```suggestion
#include <cassert> /// For std::assert
#include <iostream> /// For std::cout
#include <memory> /// For std::unique_ptr
#include <stdexcept> /// For std::out_of_range
```

**Review by realstealthninja — CHANGES_REQUESTED** [review]

File: data_structures/stack_using_array.cpp [hunk 13]
```diff
@@ -1,62 +1,176 @@
-#include <iostream>
+#include <cassert> /// For std::assert
+#include <iostream> /// For std::cout
+#include <memory> /// For std::unique_ptr
+#include <stdexcept> /// For std::out_of_range
 
-int *stack;
-int stack_idx = 0, stack_size;
+namespace data_structures {
```
Comment (realstealthninja) [inline data_structures/stack_using_array.cpp:6]:
```suggestion
/**
 * @namespace
 * @brief data_structures
 */
namespace data_structures {
```

**Review by realstealthninja — DISMISSED** [review]

**Review by realstealthninja — APPROVED** [review]
Thank you for your contribution! :heart:

**Comment by TruongNhanNguyen** [comment]
Thanks for kindly reviews, and quick responses. This PR existed for a long time, so I am delighted to see it merged.

File: data_structures/stack_using_array.cpp [hunk 14]
```diff
@@ -1,62 +1,180 @@
...
+    /**
+     * @brief Checks if the stack is full
+     *
+     * @return true if the stack is full, false otherwise
+     */
+    bool full() const { return stackIndex == stackSize - 1; }
+
+    /**
+     * @brief Checks if the stack is empty
+     *
```
Comment (realstealthninja) [inline data_structures/stack_using_array.cpp:39]:
```suggestion
```

**Review by realstealthninja — COMMENTED** [review]
Commit this the most recent push must be done by you

File: data_structures/stack_using_array.cpp (hunk 14 above)
Comment (TruongNhanNguyen) [reply data_structures/stack_using_array.cpp:39]: What do you mean by this nit?

**Review by realstealthninja — APPROVED** [review]

**Comment by realstealthninja** [comment]
Merging this! thanks for your contribution :heart:
